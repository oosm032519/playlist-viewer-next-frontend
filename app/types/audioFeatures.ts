export type Mode = "MAJOR" | "MINOR";

export interface AudioFeatures {
    acousticness: number;
    danceability: number;
    energy: number;
    instrumentalness: number;
    liveness: number;
    speechiness: number;
    valence: number;
    tempo: number;
}

export interface TrackAudioFeatures extends AudioFeatures {
    durationMs: number;
    id: string;
    key: number;
    loudness: number;
    mode: Mode;
    timeSignature: number;
}
