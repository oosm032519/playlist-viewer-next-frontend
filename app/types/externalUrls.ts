// app/types/externalUrls.ts
export interface ExternalUrls {
    externalUrls: {
        spotify: string;
    };
}
