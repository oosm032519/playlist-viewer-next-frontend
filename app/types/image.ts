// app/types/image.ts

/**
 * 画像情報。
 */
export interface Image {
    height: number;
    url: string;
    width: number;
}
