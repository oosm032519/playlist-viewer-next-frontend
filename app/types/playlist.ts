export interface PlaylistImage {
    url: string;
}

export interface Playlist {
    tracks: any;
    id: string;
    name: string;
    description: string;
    images: PlaylistImage[];
}
