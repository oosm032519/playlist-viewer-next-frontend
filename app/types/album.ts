// app/types/album.ts
import {ExternalUrls} from '@/app/types/externalUrls'
import {Image} from '@/app/types/image'

export interface Album {
    name: string;
    images: Image[];
    externalUrls: ExternalUrls;
}
