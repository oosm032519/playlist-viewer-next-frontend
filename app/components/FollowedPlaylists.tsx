// app/components/FollowedPlaylists.tsx
"use client";

import React, {useState, useEffect} from 'react';
import {Playlist} from '@/app/types/playlist';
import {Alert, AlertDescription, AlertTitle} from "./ui/alert";
import LoadingSpinner from "./LoadingSpinner";

interface FollowedPlaylistsProps {
    onPlaylistClick: (playlistId: string) => void;
}

// プレイリストデータを取得する処理をfetchFollowedPlaylists関数として抽出
const fetchFollowedPlaylists = async () => {
    try {
        const response = await fetch('/api/playlists/followed', {
            credentials: 'include'
        });
        
        if (!response.ok) {
            throw new Error('API request failed');
        }
        const data = await response.json();
        // APIレスポンスの構造に応じて適切に処理
        return Array.isArray(data) ? data : data.items || [];
    } catch (err) {
        console.error("FollowedPlaylists: エラーが発生しました", err);
        throw new Error('フォロー中のプレイリストの取得中にエラーが発生しました。');
    }
};

const FollowedPlaylists: React.FC<FollowedPlaylistsProps> = ({onPlaylistClick}) => {
    const [playlists, setPlaylists] = useState<Playlist[]>([]);
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState<string | null>(null);
    
    useEffect(() => {
        let isMounted = true;
        fetchFollowedPlaylists()
            .then(data => {
                if (isMounted) {
                    setPlaylists(data);
                }
            })
            .catch(error => {
                if (isMounted) {
                    setError(error.message);
                }
            })
            .finally(() => {
                if (isMounted) {
                    setLoading(false);
                }
            });
        
        return () => {
            isMounted = false;
        };
    }, []);
    
    if (loading) {
        return <LoadingSpinner loading={loading}/>;
    }
    if (error) {
        return (
            <Alert variant="destructive">
                <AlertTitle>Error</AlertTitle>
                <AlertDescription>{error}</AlertDescription>
            </Alert>
        );
    }
    
    return (
        <div className="mt-8">
            <h2 className="text-2xl font-bold mb-4">フォロー中のプレイリスト</h2>
            {playlists.length === 0 ? (
                <p>フォロー中のプレイリストはありません。</p>
            ) : (
                <ul className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-4">
                {playlists.map((playlist) => (
                    <li key={playlist.id} className="bg-gray-100 p-4 rounded-lg shadow">
                        <div onClick={() => onPlaylistClick(playlist.id)}>
                            {playlist.images && playlist.images.length > 0 ? (
                                <img
                                    src={playlist.images[0].url}
                                    alt={playlist.name}
                                    className="w-full h-40 object-cover mb-2 rounded-full cursor-pointer"
                                />
                            ) : (
                                <div
                                    className="w-full h-40 bg-gray-200 flex items-center justify-center mb-2 rounded-full">
                                    <span className="text-gray-500">No Image</span>
                                </div>
                            )}
                            <h3 className="font-semibold cursor-pointer">{playlist.name}</h3>
                            <p className="text-sm text-gray-600">トラック数: {playlist.tracks.total}</p>
                        </div>
                    </li>
                ))}
                </ul>
            )}
        </div>
    );
};

export default FollowedPlaylists;
