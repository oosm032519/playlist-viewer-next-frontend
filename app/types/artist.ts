// app/types/artist.ts
import {ExternalUrls} from '@/app/types/externalUrls'

export interface Artist {
    name: string;
    externalUrls: ExternalUrls;
}
