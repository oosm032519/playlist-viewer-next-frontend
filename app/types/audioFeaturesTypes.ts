// app/types/audioFeaturesTypes.ts
import {Track} from '@/app/types/track';

export interface AudioFeaturesChartProps {
    track: Track | null;
}
