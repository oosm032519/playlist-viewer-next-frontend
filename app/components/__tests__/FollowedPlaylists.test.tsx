// app/components/FollowedPlaylists.test.tsx

import React from 'react';
import {render, screen, waitFor, act, within} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import FollowedPlaylists from '../FollowedPlaylists';
import {Playlist} from '@/app/types/playlist';
import {axe, toHaveNoViolations} from 'jest-axe';

expect.extend(toHaveNoViolations);

// モックデータ
const mockPlaylists: Playlist[] = [
    {
        id: '1',
        name: 'Playlist 1',
        images: [{url: 'https://example.com/image1.jpg'}],
        tracks: {total: 10},
        description: 'This is playlist 1'
    },
    {
        id: '2',
        name: 'Playlist 2',
        images: [{url: 'https://example.com/image2.jpg'}],
        tracks: {total: 20},
        description: 'This is playlist 2'
    },
    {
        id: '3',
        name: 'Playlist 3',
        images: [],
        tracks: {total: 0},
        description: ''
    },
];

// fetchのモック
global.fetch = jest.fn(() =>
    Promise.resolve({
        ok: true,
        json: () => Promise.resolve({items: mockPlaylists}),
    } as Response)
);

// onPlaylistClickのモック関数
const mockOnPlaylistClick = jest.fn();

describe('FollowedPlaylists', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });
    
    it('renders loading spinner initially', async () => {
        // fetchのモックを遅延させる
        (global.fetch as jest.Mock).mockImplementationOnce(() =>
            new Promise(resolve => setTimeout(() => resolve({
                ok: true,
                json: () => Promise.resolve({items: mockPlaylists}),
            }), 100))
        );
        
        render(<FollowedPlaylists onPlaylistClick={mockOnPlaylistClick}/>);
        
        // LoadingSpinnerが表示されるのを待つ
        const spinner = await screen.findByRole('progressbar');
        expect(spinner).toBeInTheDocument();
        
        // プレイリストが表示されるのを待つ
        await waitFor(() => {
            expect(screen.getByText('フォロー中のプレイリスト')).toBeInTheDocument();
        });
    });
    
    it('fetches and displays playlists correctly', async () => {
        await act(async () => {
            render(<FollowedPlaylists onPlaylistClick={mockOnPlaylistClick}/>);
        });
        
        await waitFor(() => {
            expect(screen.getByText('フォロー中のプレイリスト')).toBeInTheDocument();
        });
        
        mockPlaylists.forEach((playlist) => {
            const playlistElement = screen.getByText(playlist.name);
            expect(playlistElement).toBeInTheDocument();
            
            const trackCount = screen.getByText(`トラック数: ${playlist.tracks.total}`);
            expect(trackCount).toBeInTheDocument();
            
            if (playlist.images && playlist.images.length > 0) {
                const image = screen.getByAltText(playlist.name);
                expect(image).toHaveAttribute('src', playlist.images[0].url);
            } else {
                expect(screen.getByText('No Image')).toBeInTheDocument();
            }
        });
    });
    
    it('calls onPlaylistClick when a playlist is clicked', async () => {
        render(<FollowedPlaylists onPlaylistClick={mockOnPlaylistClick}/>);
        
        // プレイリストが表示されるのを待つ
        await waitFor(() => {
            expect(screen.getByText('Playlist 1')).toBeInTheDocument();
        });
        
        // 各プレイリストをクリックしてテスト
        for (const playlist of mockPlaylists) {
            const playlistElement = screen.getByText(playlist.name);
            await userEvent.click(playlistElement);
            
            // onPlaylistClickが正しい引数で呼び出されたことを確認
            expect(mockOnPlaylistClick).toHaveBeenCalledWith(playlist.id);
            
            // モック関数をリセット
            mockOnPlaylistClick.mockClear();
        }
    });
    
    it('displays correct number of playlists', async () => {
        render(<FollowedPlaylists onPlaylistClick={mockOnPlaylistClick}/>);
        
        await waitFor(() => {
            const playlists = screen.getAllByRole('listitem');
            expect(playlists).toHaveLength(mockPlaylists.length);
        });
    });
    
    it('displays an error message when fetch fails', async () => {
        const errorMessage = 'API request failed';
        (global.fetch as jest.Mock).mockImplementationOnce(() =>
            Promise.reject(new Error(errorMessage))
        );
        
        render(<FollowedPlaylists onPlaylistClick={mockOnPlaylistClick}/>);
        
        await waitFor(() => {
            expect(screen.getByText('フォロー中のプレイリストの取得中にエラーが発生しました。')).toBeInTheDocument();
        });
    });
    
    it('handles empty playlist response', async () => {
        (global.fetch as jest.Mock).mockImplementationOnce(() =>
            Promise.resolve({
                ok: true,
                json: () => Promise.resolve([]),
            } as Response)
        );
        
        render(<FollowedPlaylists onPlaylistClick={mockOnPlaylistClick}/>);
        
        await waitFor(() => {
            expect(screen.getByText('フォロー中のプレイリスト')).toBeInTheDocument();
        });
        
        expect(screen.getByText('フォロー中のプレイリストはありません。')).toBeInTheDocument();
    });
    
    it('has no accessibility violations', async () => {
        const {container} = render(<FollowedPlaylists onPlaylistClick={mockOnPlaylistClick}/>);
        
        await waitFor(() => {
            expect(screen.getByText('フォロー中のプレイリスト')).toBeInTheDocument();
        });
        
        const results = await axe(container);
        expect(results).toHaveNoViolations();
    });
    
    it('matches snapshot', async () => {
        const {asFragment} = render(<FollowedPlaylists onPlaylistClick={mockOnPlaylistClick}/>);
        
        await waitFor(() => {
            expect(screen.getByText('フォロー中のプレイリスト')).toBeInTheDocument();
        });
        
        expect(asFragment()).toMatchSnapshot();
    });
});
