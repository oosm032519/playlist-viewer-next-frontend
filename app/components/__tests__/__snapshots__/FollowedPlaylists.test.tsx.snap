// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FollowedPlaylists matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mt-8"
  >
    <h2
      class="text-2xl font-bold mb-4"
    >
      フォロー中のプレイリスト
    </h2>
    <ul
      class="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-4"
    >
      <li
        class="bg-gray-100 p-4 rounded-lg shadow"
      >
        <div>
          <img
            alt="Playlist 1"
            class="w-full h-40 object-cover mb-2 rounded-full cursor-pointer"
            src="https://example.com/image1.jpg"
          />
          <h3
            class="font-semibold cursor-pointer"
          >
            Playlist 1
          </h3>
          <p
            class="text-sm text-gray-600"
          >
            トラック数: 10
          </p>
        </div>
      </li>
      <li
        class="bg-gray-100 p-4 rounded-lg shadow"
      >
        <div>
          <img
            alt="Playlist 2"
            class="w-full h-40 object-cover mb-2 rounded-full cursor-pointer"
            src="https://example.com/image2.jpg"
          />
          <h3
            class="font-semibold cursor-pointer"
          >
            Playlist 2
          </h3>
          <p
            class="text-sm text-gray-600"
          >
            トラック数: 20
          </p>
        </div>
      </li>
      <li
        class="bg-gray-100 p-4 rounded-lg shadow"
      >
        <div>
          <div
            class="w-full h-40 bg-gray-200 flex items-center justify-center mb-2 rounded-full"
          >
            <span
              class="text-gray-500"
            >
              No Image
            </span>
          </div>
          <h3
            class="font-semibold cursor-pointer"
          >
            Playlist 3
          </h3>
          <p
            class="text-sm text-gray-600"
          >
            トラック数: 0
          </p>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
